AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Migration/MutationSyncMetadataMigrationDelegate+SQLite.swift

```diff
@@ -0,0 +1,119 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Foundation
+import SQLite
+import AWSPluginsCore
+
+final class SQLiteMutationSyncMetadataMigrationDelegate: MutationSyncMetadataMigrationDelegate {
+
+    let modelSchemas: [ModelSchema]
+    weak var storageAdapter: SQLiteStorageEngineAdapter?
+
+    init(storageAdapter: SQLiteStorageEngineAdapter, modelSchemas: [ModelSchema]) {
+        self.storageAdapter = storageAdapter
+        self.modelSchemas = modelSchemas
+    }
+
+    func transaction(_ basicClosure: BasicThrowableClosure) throws {
+        try storageAdapter?.transaction(basicClosure)
+    }
+
+    // MARK: - Clear
+
+    func clear() throws {
+        log.debug("Clearing MutationSyncMetadata and ModelSyncMetadata to force full sync.")
+        try deleteMutationSyncMetadata()
+        try deleteModelSyncMetadata()
+    }
+
+    @discardableResult func deleteMutationSyncMetadata() throws -> String {
+        guard let storageAdapter = storageAdapter else {
+            log.debug("Missing SQLiteStorageEngineAdapter")
+            throw DataStoreError.unknown("Missing storage adapter for model migration", "", nil)
+        }
+
+        return try storageAdapter.emptyStore(for: MutationSyncMetadata.schema)
+    }
+
+    @discardableResult func deleteModelSyncMetadata() throws -> String {
+        guard let storageAdapter = storageAdapter else {
+            log.debug("Missing SQLiteStorageEngineAdapter")
+            throw DataStoreError.unknown("Missing storage adapter for model migration", "", nil)
+        }
+
+        return try storageAdapter.emptyStore(for: ModelSyncMetadata.schema)
+    }
+
+    // MARK: - Migration
+
+    func migrate() throws {
+        log.debug("Modifying and backfilling MutationSyncMetadata")
+        try dropMutationSyncMetadataCopyIfExists()
+        try createMutationSyncMetadataCopyTable()
+        try backfillMutationSyncMetadata()
+        try dropMutationSyncMetadata()
+        try renameMutationSyncMetadataCopy()
```

Comment:
it can even be a single method similar to `func applyMigrationStep(_ step: MutationSyncMetadataMigrationStep)` where `MutationSyncMetadataMigrationStep` is an enum 
```swift
enum MutationSyncMetadataMigrationStep {
  case dropMutationSyncMetadataCopy
  .....
}
```

Reply:
> all these methods listed in MutationSyncMetadataMigrationDelegate

let me give this a try and add them back to the Delegate protocol, I hastily removed them earlier

Reply:
updated the delegate to have `applyMigrationStep` and defined the steps in the enum. the delegate just passes the call to the internal method of the same name as the enum case, which may be defined/implemented different for different storage adapters